pkg/login/social/generic_oauth.go

```diff
@@ -142,20 +142,14 @@ func (s *SocialGenericOAuth) UserInfo(ctx context.Context, client *http.Client,
 			}
 		}
 
-		if userInfo.Role == "" {
-			if !s.skipOrgRoleSync {
-				role, grafanaAdmin := s.extractRoleAndAdmin(data.rawJSON, []string{}, true)
-				if role != "" {
-					s.log.Debug("Setting user info role from extracted role")
-
-					userInfo.Role = role
-					if s.allowAssignGrafanaAdmin {
-						userInfo.IsGrafanaAdmin = &grafanaAdmin
-					}
-				}
+		if (userInfo.Role == "" || userInfo.Role == s.defaultRole()) && !s.skipOrgRoleSync {
+			role, grafanaAdmin, err := s.extractRoleAndAdminOptional(data.rawJSON, []string{})
+			if err != nil {
```

Comment:
I still don't think we can return an error here - this would fail if `roleAttributeStrict` is set to `true`, but the role is being sent through the user info endpoint (it would fail, as the role won't be successfully parsed from the token).